### PR TITLE
feat: #1250 Added GenerateMetrics function, endpoint /ai/metrics, gem…

### PR DESCRIPTION
**Pull request by Moon1it**
…ini client.
https://github.com/tritonJS826/masters-way/issues/1250

### CODE REVIEW

**Review by tritonJS826 — APPROVED** [review]

File: mw-server/schemas/AI.schema.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package schemas
+
+type GenerateMetricsPayload struct {
+	WayName         string           `json:"wayName" validate:"required"`
+	GoalDescription string           `json:"goalDescription" validate:"required"`
+	Metrics         []MetricResponse `json:"metrics" validate:"required"`
+}
+
+type AIResponse struct {
+	Answer string `json:"answer"`
```
Comment (tritonJS826) [inline mw-server/schemas/AI.schema.go:10]: It shouold be a list of metrics

File: mw-server/schemas/AI.schema.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package schemas
+
+type GenerateMetricsPayload struct {
+	WayName         string           `json:"wayName" validate:"required"`
+	GoalDescription string           `json:"goalDescription" validate:"required"`
+	Metrics         []MetricResponse `json:"metrics" validate:"required"`
+}
+
+type AIResponse struct {
```
Comment (tritonJS826) [inline mw-server/schemas/AI.schema.go:9]: use more specific name

**Comment by tritonJS826** [comment]
Please, tell me how does it works for local develompent, does it use stub data to response?

**Comment by tritonJS826** [comment]
add new env variable  to app.env.example

File: mw-server/services/AI.service.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	for _, metric := range payload.Metrics {
+		prompt += fmt.Sprintf("Description: %s\n  Is Done: %t\n  Done Date: %v\n  Estimation Time: %d\n",
+			metric.Description, metric.IsDone, metric.DoneDate, metric.MetricEstimation)
+	}
+
+	return prompt
+}
+
+func GetMetricsByGoal(ctx context.Context, payload *schemas.GenerateMetricsPayload) (string, error) {
+	client, err := genai.NewClient(ctx, option.WithAPIKey(config.Env.GeminiApiKey))
```
Comment (tritonJS826) [inline mw-server/services/AI.service.go:29]: client should be reused

File: mw-server/services/AI.service.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	if err != nil {
+		return "", fmt.Errorf("failed to create client: %w", err)
+	}
+	defer client.Close()
+
+	prompt := createMetricsPrompt(payload)
+	log.Println("Generated Prompt:", prompt)
+
+	model := client.GenerativeModel("gemini-1.5-flash")
```
Comment (tritonJS826) [inline mw-server/services/AI.service.go:38]: it's kind of config, let;s move this string somewhere outside from regular code

File: mw-server/services/AI.service.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	resp, err := model.GenerateContent(ctx, genai.Text(prompt))
+	if err != nil {
+		return "", fmt.Errorf("failed to generate content: %w", err)
+	}
+
+	if len(resp.Candidates) == 0 || len(resp.Candidates[0].Content.Parts) == 0 {
+		return "", fmt.Errorf("no content generated")
+	}
```
Comment (tritonJS826) [inline mw-server/services/AI.service.go:46]: why should we handle this case separetely? Does AI returns empty strings in some cases?

File: mw-server/services/AI.service.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+	text := fmt.Sprint(resp.Candidates[0].Content.Parts[0])
+	log.Println("Generated Text:", text)
+
+	return text, nil
```
Comment (tritonJS826) [inline mw-server/services/AI.service.go:51]: could you tell me what the parts and why do we need only first one?

File: mw-server/services/AI.service.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+	if err != nil {
+		return "", fmt.Errorf("failed to generate content: %w", err)
+	}
+
+	if len(resp.Candidates) == 0 || len(resp.Candidates[0].Content.Parts) == 0 {
+		return "", fmt.Errorf("no content generated")
+	}
+
+	text := fmt.Sprint(resp.Candidates[0].Content.Parts[0])
+	log.Println("Generated Text:", text)
```
Comment (tritonJS826) [inline mw-server/services/AI.service.go:49]: we don't need to log it such way for now

File: mw-server/services/AI.service.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+	"github.com/google/generative-ai-go/genai"
+	"google.golang.org/api/option"
+)
+
+func createMetricsPrompt(payload *schemas.GenerateMetricsPayload) string {
+	var prompt string
+
+	prompt += fmt.Sprintf("User has created their own way, described the goal, and possibly made a few metrics. Based on this data, create several new metrics that align with the described goal but do not duplicate the existing ones.\n")
+	prompt += fmt.Sprintf("Give your answer as a list of Metrics.\n")
```
Comment (tritonJS826) [inline mw-server/services/AI.service.go:18]:
```suggestion
    prompt += fmt.Sprintf("Based on the provided data, create 4-6 extra metrics. Return only the list of metrics separated with comma sign with no additional comments .\n")
```

File: mw-server/services/AI.service.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+	prompt += fmt.Sprintf("Way Name: %s\nGoal Description: %s\nMetrics:\n", payload.WayName, payload.GoalDescription)
```
Comment (tritonJS826) [inline mw-server/services/AI.service.go:19]:
```suggestion
    prompt += fmt.Sprintf("Project Name: %s\nGoal : %s\nMetrics:\n", payload.WayName, payload.GoalDescription)
```

File: mw-server/services/AI.service.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+	for _, metric := range payload.Metrics {
+		prompt += fmt.Sprintf("Description: %s\n  Is Done: %t\n  Done Date: %v\n  Estimation Time: %d\n",
```
Comment (tritonJS826) [inline mw-server/services/AI.service.go:21]:
```suggestion
    prompt += fmt.Sprintf("Metrics:\n"
    prompt += fmt.Sprintf("1. (put index here): %s\n Estimation Time: %d\n"
```

File: mw-server/routes/AI.route.go [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+func NewRouteAI(aiController controllers.AIController) AIRoutes {
+	return AIRoutes{aiController}
+}
+
+func (cr *AIRoutes) AIRoute(rg *gin.RouterGroup) {
+	router := rg.Group("ai")
+	router.POST("/metrics", cr.aiController.GenerateMetrics)
+	// router.POST("/comment", cr.aiController.GenerateComment)
```
Comment (tritonJS826) [inline mw-server/routes/AI.route.go:20]: !

File: mw-server/schemas/AI.schema.go (hunk 2 above)
Comment (Moon1it) [reply mw-server/schemas/AI.schema.go:9]: Done

File: mw-server/services/AI.service.go (hunk 4 above)
Comment (Moon1it) [reply mw-server/services/AI.service.go:38]: Done

**Pull request merged**